### PR TITLE
fix(metrics): track window.onerror

### PR DESCRIPTION
**Pull request by vladikoff**
Partial fix for #1758
- [x] Needs a test
 
Requires 4 reviews
- [x] @ckarlof 
- [x] @zaach 
- [x] @kparlante 
- [x] @shane-tomlinson

### CODE REVIEW

File: app/scripts/lib/app-start.js [hunk 1]
```diff
@@ -114,25 +114,46 @@ function (
...
     },
 
+    _trackWindowOnError: function () {
+      var self = this;
+      // if startup is okay we want to log future window.onerror events
+      window.onerror = function (message /*, url, lineNumber*/) {
+        var errMsg = 'null';
+
+        if (message) {
+          errMsg = message.toString().substring(0, 100);
```
Comment (vladikoff) [inline app/scripts/lib/app-start.js]: we limit the event message to 100 characters

File: app/scripts/lib/app-start.js [hunk 2]
```diff
@@ -114,25 +114,46 @@ function (
         this.initializeInterTabChannel()
       ])
       .then(_.bind(this.allResourcesReady, this))
-      .then(null, function (err) {
-        if (console && console.error) {
-          console.error('Critical error:');
-          console.error(err);
+      .then(
+        function () {
```
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: nit: line spacing.

File: app/scripts/lib/app-start.js [hunk 3]
```diff
@@ -114,25 +114,46 @@ function (
...
         }
-        if (self._metrics) {
-          self._metrics.logError(err);
-        }
-
-        //Something terrible happened. Let's bail.
-        var redirectTo = self._getErrorPage(err);
-        self._window.location.href = redirectTo;
-      });
+      );
```
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: nit: line spacing, combine this with the previous `}`

File: app/scripts/lib/app-start.js (hunk 1 above)
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: Let's make the 100 a constant.

File: app/tests/spec/lib/app-start.js [hunk 4]
```diff
@@ -132,6 +132,19 @@ function (chai, sinon, AppStart, Session, Constants, p,
...
+      it('tracks window errors', function () {
+        return appStart.startApp()
+          .then(function () {
+            sinon.stub(appStart._metrics, 'logEvent', function () {
+            });
+
+            window.onerror.call(window, 'message', document.location.toString(), 2);
+          })
+          .then(function () {
+            assert.ok(appStart._metrics.logEvent.calledOnce);
```
Comment (shane-tomlinson) [inline app/tests/spec/lib/app-start.js]: We have a [helper](https://github.com/mozilla/fxa-content-server/blob/382c9db980753300b38b3f767aa071a1163ccf73/app/tests/lib/helpers.js#L91) that allows you to test for specific events in the event log. For an example of use, see [this snippet](https://github.com/mozilla/fxa-content-server/blob/382c9db980753300b38b3f767aa071a1163ccf73/app/tests/spec/views/sign_in.js#L204-L205).

**Comment by zaach** [comment]
:+1:

**Comment by vladikoff** [comment]
@kparlante we are looking to build a dashboard list out of these errors that would match on `error.onwindow.*` events and might look something like this:

image

**Comment by ckarlof** [comment]
yes.

**Comment by shane-tomlinson** [comment]
This is a fine piece of work, going to run the tests to make sure nothing broke.

**Comment by shane-tomlinson** [comment]
all automated tests pass, manual testing of throwing an error asynchronously in the signup view's `afterRender` function was caught and logged.

manual test code diff:
```diff
diff --git a/app/scripts/lib/app-start.js b/app/scripts/lib/app-start.js
index 5594f69..dd0b705 100644
--- a/app/scripts/lib/app-start.js
+++ b/app/scripts/lib/app-start.js
@@ -148,6 +148,8 @@ function (
         if (self._metrics) {
           self._metrics.logEvent('error.onwindow.' +  errMsg);
         }
+
+        console.error('error: %s', message);
       };
     },
 
diff --git a/app/scripts/views/sign_up.js b/app/scripts/views/sign_up.js
index 6f25ba0..14e1bee 100644
--- a/app/scripts/views/sign_up.js
+++ b/app/scripts/views/sign_up.js
@@ -60,6 +60,9 @@ function (_, p, BaseView, FormView, Template, Session, AuthErrors,
 
       this.transformLinks();
 
+      setTimeout(function () {
+        throw new Error('heya');
+      }, 1000);
       return FormView.prototype.afterRender.call(this);
     },
 
```

r+ from me!

**Comment by kparlante** [comment]
r+ looks good to me

**Pull request merged**